docs/IR.md

```diff
@@ -185,14 +185,16 @@ input|ValueInfo[]|The input “parameters” of the graph, possibly initialized
 output|ValueInfo[]|The output parameters of the graph. Once all output parameters have been written to by a graph execution, the execution is complete.
 value_info|ValueInfo[]|Used to store the type and shape information of values that are not inputs or outputs.
 
-Each main (top-level) graph MUST define the names and types of its inputs and outputs, which are specified as ‘value info’ structures, having the following properties:
+ValueInfo has the following properties:
 
 Name|Type|Description
 |---|---|---|
 name|string|The name of the value/parameter.
-type|Type|The type of the value.
+type|Type|The type of the value **including shape information**.
 doc_string|string|A human-readable documentation for this value. Markdown is allowed.
 
+Each main (top-level) graph MUST define the names, types and shapes of its inputs and outputs, which are specified as ‘value info’ structures. The main graph inputs and outputs are required to have a shape, indicating the rank, even though the exact dimensions need not be specified.
+
```

Comment:
@skottmckay , @gramalingam : Regarding the type info - it is clear from this statement that the graph inputs and outputs must have types but what about the node inputs? Are they required to have type info? 
This will determine how we handle the case of absent type info more specifically when contxt->getInputType(<input index>) is nullptr. If all inputs (graph and intermediate node inputs) require type info then the node level type inference can simply fail when context->getInputType(<input index>) is a nullptr otherwise the node level type inference should stop and return when type info is not available.

Reply:
For node inputs (that are not graph inputs), the type information is optional. The model does not *need* to specify it. But in a "correct model", type-inference will be able to infer all these types automatically. (That is to answer your question. In terms of the document: I believe this is mentioned in the document, at least it was at some point in this PR; I haven't checked the latest version.)

Reply:
I think it is just mentioned in the next sentence?
> Nested subgraphs (specified as attribute values) MUST define the names of its inputs and outputs and MAY define the types of its inputs and outputs.

Do we need to further indicate the input/output of Nested subgraphs as the node input/output here?

Another section which is talking about this:
>  When the shape property is absent for a node input, a tensor value of any shape may be passed from the caller. When the shape property is absent in the type of a value, it indicates that the corresponding runtime value may have any shape. However, types with a missing shape are not permitted in all contexts. In particular, the inputs and outputs of a model (top-level graphs) are required to have a shape, indicating the rank of inputs and outputs, even though the exact dimensions need not be specified.